### DISCUSSION TITLE
Usage without Docker

### DISCUSSION THREAD

**Discussion by runxel** [plausible/analytics#743] [Self-Hosted Support] [answered]
Any way to get this running without Docker?
My provider doesn't support Docker.

**Comment by jamiew0w** [comment]
i'm also interested in setting this up without docker. are the devs interested in offering this?

was surprised to see the issue closed and moved to discussion, but then no reply

**Reply by metmarkosaric** [reply]
we keep issues for actual bugs with the product only as explained in the issues template. all the other conversation such as feature requests like this one go into the discussions forum

**Comment by ukutaht** [comment, marked as answer]
For simplicity's sake we don't have any plans to officially support running Plausible without Docker. You can build a Mix release from source if you wish but we don't provide any instructions or help for that.

**Reply by MaluNoPeleke** [reply]
Has anyone found a way to set this up manually without Docker and wrote a documentation about it?
My hoster doesn't support Docker as well but I would like to move from Matomo to Plausible while still keeping the data on the server.

**Reply by helb** [reply]
> Has anyone found a way to set this up manually without Docker and wrote a documentation about it?

Not a documentation, but here's roughly what i did to get it running, figured it out mostly by reading Dockerfile and error messages

## Clone the repo

```bash
git clone https://github.com/plausible/analytics
cd analytics
git checkout v.1.4.4 # the latest tag right now
```

## Install DBs and Erlang compiler

- install & configure Postgres and Clickhouse to listen on localhost
- install Erlang 24, it doesn't seem to work with 25 (something about "hut" and "rebar"… i don't know shit about Erlang)


## Build frontend assets

```bash
cd assets
sed -i 's~\.\.\/deps~../app/deps~' package.json
npm install
npx browserslist@latest --update-db
npm run deploy
```

## Build tracker script

```bash
cd tracker
npm install
npm run deploy
```

## Build backend

```bash
mkdir app
cd app
export MIX_ENV=prod
wget https://s3.eu-central-1.wasabisys.com/plausible-application/geonames.csv
cp ../mix.exs  ../mix.lock ./
mix local.hex --force
mix local.rebar --force
mix deps.get --only prod
mix deps.compile
cp -r ../priv ../lib ../config ./
mix deps.get
mix phx.digest priv/static
mix download_country_database
mix sentry_recompile
mv geonames.csv priv/geonames.csv
mix release plausible
```

## Configure and run

- configure via env variables: https://plausible.io/docs/self-hosting-configuration
    ```bash
    BASE_URL="https://stats.domain.tld"
    DATABASE_URL="postgres://postgres:postgres@127.0.0.1:5432/plausible"
    SECRET_KEY_BASE="…"
    CLICKHOUSE_DATABASE_URL="http://127.0.0.1:8123/plausible_events_db"
    ADMIN_USER_EMAIL="admin@localhost"
    ADMIN_USER_PWD="asdfgh" # at least 6 characters
    ADMIN_USER_NAME="admin"
    SELFHOST="true"
    ENVIRONMENT="prod"
    MAILER_EMAIL="…"
    SMTP_HOST_ADDR="…" # it throws http 500 in several places without smtp configured
    SMTP_HOST_PORT="465"
    SMTP_USER_NAME="…"
    SMTP_USER_PWD="…"
    SMTP_HOST_SSL_ENABLED="true"
    DISABLE_REGISTRATION="true"
    ```

- populate DBs
    ```bash
    _build/prod/rel/plausible/bin/plausible eval Plausible.Release.createdb
    _build/prod/rel/plausible/bin/plausible eval Plausible.Release.migrate
    _build/prod/rel/plausible/bin/plausible eval Plausible.Release.init_admin
    _build/prod/rel/plausible/bin/plausible eval Plausible.Release.seed
    ```

- run the server to see it it works
    ```bash
    _build/prod/rel/plausible/bin/plausible start
    ```

- move the `app/_build/prod/rel/plausible` dir to some sane location
- set up supervisord to manage the service
- set up nginx reverse proxy and ssl cert

**Reply by zigomi** [reply]
@helb  What will the steps be when you need to update the source code from the origin, redeploy and relaunch it?

**Reply by zigomi** [reply]
What's this for?

```
sed -i 's~\.\.\/deps~../app/deps~' package.json
```

**Reply by helb** [reply]
> What's this for?

it changes the paths for "phoenix" and "phoenix_html" in package.json from ../deps to ../app/deps (so node can find them)

> What will the steps be when you need to update

Probably just git pull, re-run `npm run deploy` and all the mix commands, migrate db and restart. If i ever get to install it on a production server, i'll put some "real" guide together

**Reply by zigomi** [reply]
> it changes the paths for "phoenix" and "phoenix_html" in package.json from ../deps to ../app/deps (so node can find them)

But there's nothing related to "phoenix" and "phoenix_html" in package.json, is there?

**Reply by helb** [reply]
https://github.com/plausible/analytics/blob/v1.4.4/assets/package.json#L36-L37

**Comment by paulocoghi** [comment]
Same here. We are thinking in built a new open source analytics from scratch to be deployed as a single binary (and just that), since Docker is a no-go for our company (and our customers, as well other companies on our group).

**Reply by jamiew0w** [reply]
check out umami. it caught my attention just this morning (which really is a mind bending coincidence considering i posted in this thread a year ago) and here you are, same question.

**Reply by paulocoghi** [reply]
Thank you a lot, @jamiew0w 

I also found Ackee (https://github.com/electerious/Ackee -- https://ackee.electerious.com/) which, as Unami, can be installed without Docker.

It's sad, since Plausible is a very mature option, but a stack that forces Docker is, unfortunately, a negative aspect.

**Reply by paulocoghi** [reply]
A negative point of Unami and Ackee is the NodeJS runtime as pre-requisite. But it is (still) way better than having to use Docker.

Month after month of research makes me more inclined to develop a new alternative with a single and portable binary, made on a compiled language, without the need of runtime libraries nor Docker.

**Reply by tamirciat** [reply]
also check this one https://www.npmjs.com/package/plausible-tracker

**Reply by nakoo** [reply]
If you're using Nix OS, you can run it natively. 
https://github.com/NixOS/nixpkgs/blob/nixos-21.11/nixos/modules/services/web-apps/plausible.nix

**Reply by paulocoghi** [reply]
Thanks, @nakoo, but since we and our customers need an OS-agnostic single-binary solution, we are inclined to develop a new project from scratch.

**Reply by zigomi** [reply]
> .....since Docker is a no-go for our company (and our customers, as well other companies on our group).

@paulocoghi why not?

**Reply by paulocoghi** [reply]
On many of the articles below you will see that Docker is slow, complex, heavy and insecure [1][2][3][...], which is the opposite of what we and our customers want as a pre-requisite for a software that must be easily installed, secure and lightweight.

And our practical experience has shown the same.

[1] https://www.freecodecamp.org/news/7-cases-when-not-to-use-docker/
[2] https://www.howtogeek.com/devops/when-not-to-use-docker-cases-where-containers-dont-help/
[3] https://hackaday.com/2019/06/07/this-week-in-security-nvidia-ransomware-retirement-and-a-toctou-bug-in-docker/
[4] https://techbeacon.com/security/hackers-love-docker-container-catastrophe-3-2-1
[5] https://www.cimcor.com/blog/the-top-5-security-risks-in-docker-container-deployment
[6] https://www.reddit.com/r/docker/comments/m5092k/why_we_dont_use_docker_we_dont_need_it/
[7] https://stackoverflow.com/questions/24394243/why-are-docker-container-images-so-large
